integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/columnLineage/OutputFieldsCollector.java

```diff
@@ -0,0 +1,48 @@
+package io.openlineage.spark3.agent.lifecycle.plan.columnLineage;
+
+import io.openlineage.spark.agent.util.ScalaConversionUtils;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.spark.sql.catalyst.expressions.Attribute;
+import org.apache.spark.sql.catalyst.expressions.NamedExpression;
+import org.apache.spark.sql.catalyst.plans.logical.Aggregate;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import org.apache.spark.sql.catalyst.plans.logical.Project;
+
+/** Class created to collect output fields with the corresponding ExprId from LogicalPlan. */
+class OutputFieldsCollector {
+
+  private final LogicalPlan plan;
+
+  OutputFieldsCollector(LogicalPlan plan) {
+    this.plan = plan;
+  }
+
+  void collect(ColumnLevelLineageBuilder builder) {
+    collect(plan, builder);
+  }
+
+  private void collect(LogicalPlan plan, ColumnLevelLineageBuilder builder) {
+    List<NamedExpression> expressions =
+        ScalaConversionUtils.fromSeq(plan.output()).stream()
+            .filter(attr -> attr instanceof Attribute)
+            .map(attr -> (Attribute) attr)
+            .collect(Collectors.toList());
+
+    if (plan instanceof Aggregate) {
+      expressions.addAll(
+          ScalaConversionUtils.<NamedExpression>fromSeq(((Aggregate) plan).aggregateExpressions()));
+    } else if (plan instanceof Project) {
+      expressions.addAll(
+          ScalaConversionUtils.<NamedExpression>fromSeq(((Project) plan).projectList()));
+    }
```

Comment:
Nope. At the bottom of the method, we call it recursively on its children if the the output is not discovered.

It would have missed output nodes for `SaveIntoDataSourceCommand` because in case of this class `query` field is not returned with `children` but `SaveIntoDataSourceCommand` is no longer available in Spark 3.x.